corehq/motech/repeaters/tasks.py

```diff
@@ -66,15 +69,36 @@ def clean_logs():
     queue=settings.CELERY_PERIODIC_QUEUE,
 )
 def check_repeaters():
+    # this creates a task for all partitions
+    # the Nth child task determines if a lock is available for the Nth partition
+    for current_partition in range(CHECK_REPEATERS_PARTITION_COUNT):
+        check_repeaters_in_partition.delay(current_partition, CHECK_REPEATERS_PARTITION_COUNT)
+
+
+def iterate_repeat_record_ids_for_partition(start, partition, total_partitions):
+    for record_id in iterate_repeat_record_ids(start, chunk_size=100000):
+        if hash(record_id) % total_partitions == partition:
```

Comment:
rather than going through the full list why not use `limit` and `skip` to only fetch a certain portion of the view?

Reply:
We iterate through the entire list to retrieve all of the doc ids that are part of the current partition being processed. I'm not sure if you're suggesting to assign a chunk to each partition (using skip and limit) instead of filtering ids down. If you are then the reason for the current implementation is to consistently group the same doc ids into the same partitions, and correctly check for new docs the next time `check_repeaters` is called. Others may be able to provide a better explanation.

This may not be what you are suggesting so feel free to clarify more.

Reply:
The main issue I have is that each of the 16 partitions is going to have to iterate through the entire list which seems wasteful.

> I'm not sure if you're suggesting to assign a chunk to each partition (using skip and limit) instead of filtering ids down.

Yes that's what I'm suggesting.

> consistently group the same doc ids into the same partitions

I'm curious why this is necessary. I don't see any reason for this to be a requirement.

Is the concern that using skip/limit will result in inconsistent results (some rows missing?). If that's the case you can use the same technique as [paginate_view](https://github.com/dimagi/commcare-hq/blob/1405cb2124e4ab60f7c5d4358b35fa10bb5227a4/corehq/util/couch_helpers.py#L151-L153) and effectively chain multiple tasks together:

```python
def process_repeat_records(num_to_process, start_from_doc_id=None):
    doc_ids = []
    doc_ids = [doc_id in iterate_doc_ids(num_to_process, start_from_doc_id)]
    if not doc_ids:
        return

    # assume there are more docs to process so spawn another task
    process_repeat_records.delay(num_to_process, start_from_doc_id=doc_ids[-1])

    # process the docs in doc_ids
```

Based on your timing data above this may all be moot (30s to iterate over the whole view isn't that crazy).

Reply:
Ah that makes sense. The previous implementation fetches all docs, so it is an improvement to fetch all doc ids instead, then filter down to only fetch a chunk of docs. I'm curious to hear what others think about your proposal because that does appear more efficient than the current implementation. 

>>>I'm curious why this is necessary. I don't see any reason for this to be a requirement.

I'm having a hard time justifying why it's necessary myself. I'll let others speak to this. 

>>>Based on your timing data above this may all be moot (30s to iterate over the whole view isn't that crazy).

It is also worth noting because this change will allow us to process repeat records faster (using multiple celery workers), that time should drop more since this queue is pretty backed up right now (I think).

Reply:
I am not sure I understand how `limit` and `skip` can be parallelized the same way. This is certainly not the most efficient implementation, but it allows us to have 16 processes at once and not worry about them trying to process the same docs (since what this does is spin off celery tasks, and reprocessing the same docs means clogging the celery queue with duplicates). You can't coordinate pagination like that since the underlying view is constantly changing.

It only takes 30s to loop over all the ids but that is without even fetching the docs, and thats not necessarily the slow part of this task. It also modifies and saves the results. And the current version, which does it all at once, takes over 20 hours.